### PR TITLE
Issue #208 ( $map` on objects is weird when `each` results on an array)

### PR DESCRIPTION
**Pull request by Biboswan**
Added a test case .
Throwing an error if each(x)  does not  compute to an object.

### CODE REVIEW

File: specification.yml [hunk 1]
```diff
@@ -309,6 +309,11 @@ context:  {}
 template: {$$map: [2, 4, 6], each(x): {$$eval: "x + a"}}
 result:   {$map: [2, 4, 6], each(x): {$eval: "x + a"}}
 ---
+title:	  escape $map as object and each(identifier) as array
+context:  {}
+template: {$map: {"hello": 5, "test": 9},"each(x)":['${x.key}',{$eval: 'x.val+1'}]}
+error:   'TemplateError: $map is an object so each(identifier) shall also compute to an object'
```
Comment (jonasfj) [inline specification.yml]: we can do a better error message I'm sure :)

File: src/builtins.js [hunk 2]
```diff
@@ -62,7 +62,7 @@ module.exports = (context) => {
   });
 
   ['sqrt', 'ceil', 'floor', 'abs'].forEach(name => {
-    if (Math[name] == undefined) {
+    if (Math[name] == undefined) {
```
Comment (jonasfj) [inline src/builtins.js]: ?

File: src/index.js [hunk 3]
```diff
@@ -176,6 +176,10 @@ operators.$map = (template, context) => {
 
   if (object) {
     value = Object.keys(value).map(key => ({key, val: value[key]}));
+    let firstRender= render(each, Object.assign({}, context, {[x]: value[0]}));
+    if  (!isObject(firstRender)) {
+      throw new TemplateError('$map is an object so each(identifier) shall also compute to an object');
+    }
   }
```
Comment (jonasfj) [inline src/index.js]: Have a look at the golang implementation... this is hard to read... I propose refactoring the entire method into:
```js
if (isObject(value)) {
  ....
  return ...;
} else {
  ....
  return ...;
}
```

even if that means we have to have a little code duplication, I think this will be prettier in the long run.

File: src/index.js [hunk 4]
```diff
@@ -176,6 +176,10 @@ operators.$map = (template, context) => {
 
   if (object) {
     value = Object.keys(value).map(key => ({key, val: value[key]}));
+    let firstRender= render(each, Object.assign({}, context, {[x]: value[0]}));
```
Comment (jonasfj) [inline src/index.js]: Also `yarn run lint` will run a linter on the code... it's similar to `yarn test`.. :)

File: src/builtins.js (hunk 2 above)
Comment (Biboswan) [reply src/builtins.js]: I just wrote a comment earlier which I just removed now

File: specification.yml (hunk 1 above)
Comment (Biboswan) [reply specification.yml]: yes.:)Invalid input..

File: src/index.js (hunk 4 above)
Comment (Biboswan) [reply src/index.js]: Yeah it ran  successfully

Comment (djmitche) [reply src/index.js]: It should have flagged the trailing whitespace on line 65 and the missing ` ` on this line..

File: specification.yml [hunk 5]
```diff
@@ -309,6 +309,11 @@ context:  {}
 template: {$$map: [2, 4, 6], each(x): {$$eval: "x + a"}}
 result:   {$map: [2, 4, 6], each(x): {$eval: "x + a"}}
 ---
+title:    escape $map as object and each(identifier) as array
```
Comment (djmitche) [inline specification.yml]: This test isn't for escaping (that's the `$$`).  So, good test, but should be moved to the section with all of the other $map tests.

File: src/index.js [hunk 6]
```diff
@@ -176,14 +176,16 @@ operators.$map = (template, context) => {
 
   if (object) {
     value = Object.keys(value).map(key => ({key, val: value[key]}));
-  }
-
-  value = value.map(v => render(each, Object.assign({}, context, {[x]: v})))
-    .filter(v => v !== deleteMarker);
-
-  if (object) {
+    let FirstEachValue = render(each, Object.assign({}, context, {[x]: value[0]}));
```
Comment (djmitche) [inline src/index.js]: Variable names should be interCaps without an initial capital letter (so `firstEachValue`)

File: src/index.js [hunk 7]
```diff
@@ -176,14 +176,16 @@ operators.$map = (template, context) => {
...
+    if  (!isObject(FirstEachValue)) {
+      throw new TemplateError(`$map on objects expects each(${x}) to evaluate to an object`);
+    }
+    value = value.map(v => render(each, Object.assign({}, context, {[x]: v})))
```
Comment (djmitche) [inline src/index.js]: Hm, this ends up rendering the first value twice -- which is almost OK but may be surprising if someone is using a custom function that's called more of then than expected.  But more importantly, it only checks the type of the first value.  Why not check them all in this loop?

Comment (Biboswan) [reply src/index.js]: Then shall generator function be used here like in python implementation.

Comment (djmitche) [reply src/index.js]: I don't think you need a generator, but `=>` is already defining a function, so it just needs an additional statement in that function.

File: README.md [hunk 8]
```diff
@@ -303,18 +301,33 @@ context:  {a: 1}
...
 ```
 
+The array or object is the value of the `$map` property, and the expression to evaluate
+is given by `each(var)` where `var` is the name of the variable containing each
+element. In the case of iterating over an object, `var` will be an object with two keys:
+`key` and `val`. These keys correspond to a key in the object and its corresponding value.
+
+When given an object, map returns an object only when the expression defined by `each(var)` 
+evaluates to an object since results of each 'each(var)' are merged internally to give the 
+resulting object.Otherwise the expression becomes invalid for the $map operator.
```
Comment (djmitche) [inline README.md]: Missing ` ` between `.` and `O`

File: README.md [hunk 9]
```diff
@@ -303,18 +301,33 @@ context:  {a: 1}
...
 ```
+ 
+```yaml
+template: 
+  $map: {"hello": 5, "test": 9}
+  each(x):['${x.key}',{$eval: 'x.val+1'}]
+context: {}
+error:   'TemplateError: $map on objects expects each(x) to evaluate to an object'
+```
+
```
Comment (djmitche) [inline README.md]: I don't think we need to include error cases in the README..

Comment (Biboswan) [reply README.md]: I also think so.

File: README.md (hunk 8 above)
Comment (djmitche) [reply README.md]: @Biboswan can you push a fix for this, too, while we wait for @jonasfj's feedback?

Comment (Biboswan) [reply README.md]: Didn't get, missing what??

Comment (Biboswan) [reply README.md]: ok got it space?

File: README.md [hunk 10]
```diff
@@ -303,6 +301,16 @@ context:  {a: 1}
 result:   [3, 5, 7]
 ```
 
+The array or object is the value of the `$map` property, and the expression to evaluate
+is given by `each(var)` where `var` is the name of the variable containing each
+element. In the case of iterating over an object, `var` will be an object with two keys:
+`key` and `val`. These keys correspond to a key in the object and its corresponding value.
+
+When given an object, map returns an object only when the expression defined by `each(var)` 
+evaluates to an object since results of each 'each(var)' are merged internally to give the
```
Comment (jonasfj) [inline README.md]: nit: but this can be written more precisely...

'only when' is not a good way to describe the legal constructs...

Something like:
> When `$map` is given an object, the `each(var)` property must evaluate to an object for each key/value pair `var = {key, val}` in the object given. The objects constructed by by `each(var)` will then be merged together, with later keys overwriting previous keys.

File: jsone/render.py [hunk 11]
```diff
@@ -183,21 +183,22 @@ def map(template, context):
...
         for elt in value:
             subcontext[each_var] = elt
             elt = renderValue(each_template, subcontext)
             if elt is not DeleteMarker:
                 yield elt
-
     if is_obj:
+        value = [{'key': v[0], 'val': v[1]} for v in value.items()]
         v = dict()
         for e in gen():
```
Comment (jonasfj) [inline jsone/render.py]: is `gen` referencing a variable that is declared very early in this function and then changed between the definition of `gen` and the usage of `gen`? That seems unfortunate and very hard to follow.

either pass value to `gen` as a parameter, or something else... I think passing as parameter would make a lot of sense in this case...

Note: this code actually works, but if you analyze why, it's because `value = ` is declared before `gen`... it's very hard to follow.

File: jsone/render.py [hunk 12]
```diff
@@ -183,21 +183,22 @@ def map(template, context):
...
-            v.update(e)
+            if not isinstance(e, dict):
+                raise TemplateError(
+                    "$map on objects expects {0} to evaluate to an object".format(each_key))
+            else:
```
Comment (jonasfj) [inline jsone/render.py]: `else` is not needed here...
If we `raise` an exception then that's is going to break execution and return from the function...

File: src/index.js [hunk 13]
```diff
@@ -173,17 +173,22 @@ operators.$map = (template, context) => {
...
-  value = value.map(v => render(each, Object.assign({}, context, {[x]: v})))
-    .filter(v => v !== deleteMarker);
-
-  if (object) {
+    let eachValue;
+    value = value.map(v => {
+      eachValue = render(each, Object.assign({}, context, {[x]: v}));
+      if  (!isObject(eachValue)) {
+        throw new TemplateError(`$map on objects expects each(${x}) to evaluate to an object`);
+      } else {
```
Comment (jonasfj) [inline src/index.js]: again the `else` is not needed here... if we throw an exception is going to break execution of the function

File: src/index.js [hunk 14]
```diff
@@ -173,17 +173,22 @@ operators.$map = (template, context) => {
...
+        return eachValue;
+      }
+    }).filter(v => v !== deleteMarker);
     return value.reduce((a, o) => Object.assign(a, o), {});
```
Comment (jonasfj) [inline src/index.js]: I think we can do: `return Object.assign({}, ...value);`, granted I might be wrong...

File: src/index.js [hunk 15]
```diff
@@ -173,17 +173,22 @@ operators.$map = (template, context) => {
...
   } else {
+    value = value.map(v => render(each, Object.assign({}, context, {[x]: v})))
+      .filter(v => v !== deleteMarker);
     return value;
```
Comment (jonasfj) [inline src/index.js]: Just do a:
```js
return value.map(
  v => render(....)
).filter(....);
```
or something like that...

Writing `value = value` followed by `return value` doesn't make it easier to read...

**Comment by Biboswan** [comment]
@ jonasfj  now its fine I think

**Comment by djmitche** [comment]
Awesome!
